### PR TITLE
Add "advice" and Error handlers

### LINKED ISSUE #9: Global error handler/error handling middleware
Right now if the program dies, it dies. We should be able to catch these errors and handle them based on how the user of the library wants. Note the internal code will not die, but rather in the handlers of the defined routes. Ideally the errors in routing code (ie. userspace code), will be caught and translated into a 500 of some kind.

### LINKED ISSUE #20: End-To-End testing
We should probably setup some end-to-end tests!

### PR DESCRIPTION
**Pull request by rawleyfowler**
### Advice

Advice function as end of stack middleware, they take Responses and return Responses.

middleware -> actual handler -> advice

Declare them like:
```raku
advice(-> $response { say $response; $response });
```

### Error Handlers

Error handlers can be added by specific type of error, they will catch any exception thrown of type error provided and pass the error as a parameter to the handler.

Too catch all `X::AdHoc` you would do:
```raku
error(X::AdHoc, -> $exn { Response.new(status => Http::Status(500) });
```

### CODE REVIEW

**Comment by rawleyfowler** [comment]
@masukomi Would you mind looking at this?

File: lib/Humming-Bird/Core.rakumod [hunk 1]
```diff
@@ -259,20 +259,20 @@ class Response is HTTPAction is export {
     }
 
     # Write a string to the body of the response, optionally provide a content type
-    method write(Str $body, Str $content_type = 'text/plain', --> Response) {
+    method write(Str:D $body, Str:D $content_type = 'text/plain', --> Response) {
```
Comment (2colours) [inline lib/Humming-Bird/Core.rakumod:262]: This isn't a modification but that comma in front of the return arrow looks strange to me. I wonder what purpose it serves or how it works.

Comment (rawleyfowler) [reply lib/Humming-Bird/Core.rakumod:262]: It doesn't actually do anything. I just like it as a separator on optional args and return type.

File: lib/Humming-Bird/Core.rakumod [hunk 2]
```diff
@@ -300,15 +300,14 @@ my constant $PARAM_IDX = ':';
 class Route is Callable {
     has Str $.path;
     has &.callback;
+    has &.composition;
     has @.middlewares is Array; # List of functions that type Request --> Request
 
     method CALL-ME(Request:D $req) {
         my $res = Response.new(status => HTTP::Status(200));
         if @!middlewares.elems {
```
Comment (2colours) [inline lib/Humming-Bird/Core.rakumod:307]: Again... didn't actually change but be aware of how lazy arrays could interfere with the logic: one can't call `.elems` on them and therefore they would all turn into `False` with regards to the `if`. A plain Bool check might perform better (that tries to reify exactly one element to figure out if the list is empty or not) if you need to support laziness.

File: lib/Humming-Bird/Core.rakumod [hunk 3]
```diff
@@ -357,9 +359,9 @@ sub dispatch_request(Request $request --> Response) {
     for @uri_parts -> $uri {
         my $possible_param = %loc.keys.first: *.starts-with($PARAM_IDX);
 
-        if (not %loc{$uri}:exists) && (not $possible_param) {
+        if (not %loc{$uri}:exists) && !$possible_param {
```
Comment (2colours) [inline lib/Humming-Bird/Core.rakumod]: To negate :exists, you can also do :!exists. 🤓

File: lib/Humming-Bird/Core.rakumod [hunk 4]
```diff
@@ -408,23 +420,49 @@ sub group(@routes, @middlewares) is export {
     .(@middlewares) for @routes;
 }
 
-sub routes(--> Hash) is export {
+multi sub advice(--> List:D) is export {
+    @ADVICE.clone;
+}
+
+multi sub advice(@advice) is export {
+    @ADVICE.push: |@advice;
```
Comment (2colours) [inline lib/Humming-Bird/Core.rakumod]: If I remember correctly, this is exactly what `@ADVICE.append: @advice` would do. https://new-raku.finanalyst.org/type/Array.html#method_append

File: lib/Humming-Bird/Core.rakumod [hunk 5]
```diff
@@ -408,23 +420,49 @@ sub group(@routes, @middlewares) is export {
...
-        my Request $request = Request.encode($raw_request);
-        my Bool $keep_alive = False;
-        with $request.headers<Connection> {
-            $keep_alive = $request.headers<Connection>.lc eq 'keep-alive';
+sub handle($raw-request) {
+    my Request $request = Request.encode($raw-request);
+    my Bool $keep-alive = False;
+    my &advice = [o] @ADVICE; # Advice are Response --> Response
+    with $request.headers<Connection> {
+        $keep-alive = $request.headers<Connection>.lc eq 'keep-alive';
```
Comment (2colours) [inline lib/Humming-Bird/Core.rakumod]: `with` sets the topic so I'd consider not typing `$request.headers<Connection>` again.

**Pull request merged**